### PR TITLE
Moar grover

### LINKED ISSUE #88: Add class method to run an instance of Grover's
It would be nice to match the API of Grover's to that of Simon's and Bernstein-Vazirani.

### PR DESCRIPTION
**Pull request by ampolloreno**
Fixes #88.
Also note that `[, ]` are not allowed in Quil, so we've changed the autogenerated defgate labels to use dashes.

### CODE REVIEW

File: grove/amplification/grover.py [hunk 1]
```diff
@@ -20,26 +20,31 @@
 import pyquil.quil as pq
 from pyquil.gates import H
 from pyquil.quilbase import Qubit
+from pyquil.api import JobConnection
```
Comment (jotterbach) [inline grove/amplification/grover.py]: where is this import used?

File: grove/amplification/grover.py [hunk 2]
```diff
@@ -61,41 +66,51 @@ def _construct_grover_circuit(self):
...
+        oracle.inst(tuple([oracle_name] + self.qubits))
+        self.grover_circuit = self.oracle_grover(oracle, self.qubits)
 
     def _init_attr(self, bitstring_map):
         """Initializes an instance of Grover's Algorithm given a bitstring_map.
 
-        :param dict bitstring_map: dict with string keys corresponding to bitstrings, and integer
+        :param bitstring_map: dict with string keys corresponding to bitstrings, and integer
          values corresponding to the desired phase on the output state.
+        :type bitstring_map: dict[str, int]
```
Comment (jotterbach) [inline grove/amplification/grover.py]: why do we suddenly change the formatting of how we annotate types?

File: grove/amplification/grover.py [hunk 3]
```diff
@@ -61,41 +66,51 @@ def _construct_grover_circuit(self):
...
-        :return: The corresponding instance of Grover's Algorithm.
-        :rtype: Program
+        In particular, this will prepare an initial state in the uniform superposition over all bit-
+        strings, an then use Grover's Algorithm to pick out the desired bitstring.
+
+        :param JobConnection cxn: the connection to the Rigetti cloud to run pyQuil programs.
+        :param bitstring_map: a mapping from bitstrings to the phases that the
+         oracle should impart on them. If the oracle should "look" for a bitstring, it should have a
+         ``-1``, otherwise it should have a ``1``.
+        :type bitstring_map: dict[str, int]
```
Comment (jotterbach) [inline grove/amplification/grover.py]: ditto

File: grove/amplification/grover.py (hunk 1 above)
Comment (ampolloreno) [reply grove/amplification/grover.py]: interesting, nowhere. Pycharm didn't tell me. :(

File: grove/amplification/grover.py (hunk 2 above)
Comment (ampolloreno) [reply grove/amplification/grover.py]: @jotterbach So I had a lot (not all) of them written in the python 3 typing format how you had them, but that doesn't link with the type in the documentation. It doesn't make sense to me that when you document the type in the docstring you give the type in the `typing` module.

File: grove/amplification/grover.py (hunk 3 above)
Comment (ampolloreno) [reply grove/amplification/grover.py]: Same as above.

**Comment by ampolloreno** [comment]
@jotterbach I'll add an example notebook later.

File: grove/amplification/grover.py (hunk 2 above)
Comment (jotterbach) [reply grove/amplification/grover.py]: I agree, but let's not just randomly start to change this in select places. If we change it we should at least do this module wide, better even across the full code-base. I'd like to keep consistency.

Comment (ampolloreno) [reply grove/amplification/grover.py]: @jotterbach That's fine. Is it consistent right now? Part of the reason I didn't feel strongly against changing is I thought it wasn't consistent.

Comment (ampolloreno) [reply grove/amplification/grover.py]: In any case, moving forward I think this is the right way to do it- would you rather I change them back and do a module/package-wide update later?

Comment (jotterbach) [reply grove/amplification/grover.py]: I'd prefer that, yes, because that makes the changes more atomic and contained into dedicated units.

Comment (ampolloreno) [reply grove/amplification/grover.py]: aye aye

Comment (ampolloreno) [reply grove/amplification/grover.py]: done

**Pull request merged**